### ISSUE TITLE
network dead?

### ISSUE THREAD

**Issue by RazerStvH** [PurpleI2P/i2pd#2057]
Is everyone's network dead? I got this on the server all day:
Tunnel creation success rate: 1%
Routers: 25462 Floodfills: 1518 LeaseSets: 556
Client Tunnels: 8 Transit Tunnels: 19881

**Comment by Vort** [comment]
Network is under attack right now.
1 hop tunnels work relatively well, 3 hop tunnels are almost unusable.

**Comment by Vort** [comment]
I suggest to wait for commits with mitigations and to build new binary from sources then.

**Comment by Vort** [comment]
Please try 6ce2c30. Protection should activate after 10 minutes of uptime.
For some OSes binaries are available for download from Actions page.
For other - should be built manually.

**Comment by SergSel2006** [comment]
I heared that java I2P routers are not affected by this. I tested the java router and it worked fine. Something is wrong with I2Pd

**Comment by lcharles123** [comment]
> Please try [6ce2c30](https://github.com/PurpleI2P/i2pd/commit/6ce2c30522a01534521354e83ee94f24d17bbf6e). Protection should activate after 10 minutes of uptime. For some OSes binaries are available for download from Actions page. For other - should be built manually.

Im using this new build, working well for now. Thanks!

Uptime: 10 hours, 8 minutes, 0 seconds
Network status: Firewalled - Symmetric NAT
Tunnel creation success rate: 2%
Received: 583.72 MiB (9.41 KiB/s)
Sent: 357.96 MiB (6.60 KiB/s)
Transit: 29.15 MiB (0.00 KiB/s)
Data path: /root/.i2pd
Hidden content. Press on text to see.
Routers: 5502 Floodfills: 932 LeaseSets: 0
Client Tunnels: 10 Transit Tunnels: 2

**Comment by orignal** [comment]
> I heared that java I2P routers are not affected by this. I tested the java router and it worked fine.

Because they ban all clones including real routers. I wouldn't call it "fine".

**Comment by twofaktor** [comment]
Bitcoin network affected too:

```
Apr 23 13:47:50 bitcoind[242178]: 2024-04-23T13:47:50Z [i2p] Creating persistent SAM session a274793822 with 127.0.0.1:7656
Apr 23 13:47:50 bitcoind[242178]: 2024-04-23T13:47:50Z [i2p] Error listening: Receive timeout (received 0 bytes without terminator before that)
```

**Comment by SergSel2006** [comment]
> Bitcoin network affected too:
> 
> ```
> Apr 23 13:47:50 bitcoind[242178]: 2024-04-23T13:47:50Z [i2p] Creating persistent SAM session a274793822 with 127.0.0.1:7656
> Apr 23 13:47:50 bitcoind[242178]: 2024-04-23T13:47:50Z [i2p] Error listening: Receive timeout (received 0 bytes without terminator before that)
> ```

Obviously. Now anything that relates on i2p tunnels won't work because, well, there is no i2p tunnels.

**Comment by SilmorSenedlen** [comment]
> Network is under attack right now.

Just got OOB on Windows host:
![I2Pd_DoS_New](https://github.com/PurpleI2P/i2pd/assets/6761903/be0811ee-d9d6-4655-a7d4-3f4b63cfe950)
I shooted this in process of killing .
When I saw that my browser had crashed, i2pd consume ~80 GiB of virtual memory.

**Comment by Vort** [comment]
> When I saw that my browser had crashed, i2pd consume ~80 GiB of virtual memory.

I noticed similar problem with my node.

> Maybe this is also a problem with memory leaks, and not only because of the number of requests from the network.

Attack triggers some (still unknown) bug in i2pd.
Not directly, but because of high load.

In my case leak started to happen after computer was overloaded.
Before that consumption was realtively stable at ~500MB.

**Comment by SilmorSenedlen** [comment]
> Before that consumption was relatively stable at ~500MB.

Yep,my node usually fits the same mem boundaries.
The stable load for the last month has been ~5-6 MiB and ~30-40K transit tunnels.

**Comment by LLE8** [comment]
In my experience, while normal operation (not under DDoS-attack) memory leak is from 500KB to 1MB per day, but not every day.

**Comment by Vort** [comment]
> In my experience, while normal operation (not under DDoS-attack) assumed memory leak is from 500KB to 1MB per day, but not every day.

This is, most likely, how heap fragmentation _(problem common for many C++ programs)_ manifests itself.

In comparison, for two cases, which I observed recently, leak happened with speeds of ~1MB/s and ~1GB/s.

**Comment by SilmorSenedlen** [comment]
Now I have found that the process has bloated again and consumes ~14 GiB of memory.
![i2pd_Res_02](https://github.com/PurpleI2P/i2pd/assets/6761903/12ed9389-9abb-433b-9250-470794d07367)
![i2pd_Res_01](https://github.com/PurpleI2P/i2pd/assets/6761903/530ecd4c-f84d-4e84-9e85-a3aa3de42daf)
While I was looking at the virtual memory graph, it remained at the same level, although the Working Set grew by 4-8 KiB every few seconds.
Now I see that Virtual Memory has grown in one leap.

**Comment by Vort** [comment]
I figured out that one of memory leaks (fast, happens during shutdown) appears at this line:
https://github.com/PurpleI2P/i2pd/blob/d3b699d7cdec72e6513d7a55c711e6e21902dba4/libi2pd/Transports.cpp#L474
`m_Peers` there have size of `0xfffffffffffffff7` and it starts to call `new` function with 2GB blocks.

It's in fact the same problem as in #2058 - `m_Peers` needs more mutex locks.

**Comment by SilmorSenedlen** [comment]
Looking at the memory consumption at this moment at first I thought that the app was crashed and restarted. But it turned out that ~ 15-16 GiB, which I observed during the day, collapsed at some point x_X
Although the amount of data being passed thought router  and the number of tunnels has not decreased at all.
Are these some kind of "temporary" memory leaks that the application can fix itself over time?
<details>

![i2pd_Res_03](https://github.com/PurpleI2P/i2pd/assets/6761903/3e20ed4f-35db-463d-817d-578436a27dc3)
![i2pd_Res_04](https://github.com/PurpleI2P/i2pd/assets/6761903/323bb9a0-3fa6-44fb-a62c-80cac2c0809b)

</details>

**Comment by SergSel2006** [comment]
> Looking at the memory consumption at this moment, at first I thought that the app was crashed and restarted. But it turned out that ~15-16 GiB, which I observed during the day, collapsed at some point x_X Although the amount of data being passed thought router and the number of tunnels has not decreased at all. Are these some kind of "temporary" memory leaks that app can fix itself over time?
There is no such thing as garbarage collector in C++, all memory in controlled by application. It feels like there was something with memory leak, but then that something was destroyed, freeing all leaked memory back.

**Comment by orignal** [comment]
It's a race condition that damages std structures.
